src/CPUActivityAgent.cpp

```diff
@@ -0,0 +1,532 @@
+            m_core_scal.push_back({m_platform_io.push_signal("MSR::CPU_SCALABILITY_RATIO",
+                                                             GEOPM_DOMAIN_CORE,
+                                                             domain_idx), NAN});
+            m_core_freq_control.push_back({m_platform_io.push_control("CPU_FREQUENCY_MAX_CONTROL",
+                                                                      GEOPM_DOMAIN_CORE,
+                                                                      domain_idx), -1});
+        }
+
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+            m_qm_rate.push_back({m_platform_io.push_signal("MSR::QM_CTR_SCALED_RATE",
+                                                           GEOPM_DOMAIN_PACKAGE,
+                                                           domain_idx), NAN});
+
+            m_uncore_freq_status.push_back({m_platform_io.push_signal("CPU_UNCORE_FREQUENCY_STATUS",
+                                                                      GEOPM_DOMAIN_PACKAGE,
+                                                                      domain_idx), NAN});
+        }
+
+        for (int domain_idx = 0; domain_idx < M_NUM_PACKAGE; ++domain_idx) {
+            m_uncore_freq_min_control.push_back({m_platform_io.push_control("CPU_UNCORE_FREQUENCY_MIN_CONTROL",
+                                              GEOPM_DOMAIN_PACKAGE,
+                                              domain_idx), -1});
+            m_uncore_freq_max_control.push_back({m_platform_io.push_control("CPU_UNCORE_FREQUENCY_MAX_CONTROL",
+                                              GEOPM_DOMAIN_PACKAGE,
+                                              domain_idx), -1});
+        }
+
+        //Configuration of QM_CTR must match QM_CTR config used for training data
+        m_platform_io.write_control("MSR::PQR_ASSOC:RMID", GEOPM_DOMAIN_BOARD, 0, 0);
+        m_platform_io.write_control("MSR::QM_EVTSEL:RMID", GEOPM_DOMAIN_BOARD, 0, 0);
+        m_platform_io.write_control("MSR::QM_EVTSEL:EVENT_ID", GEOPM_DOMAIN_BOARD, 0, 2);
+    }
+
+    // Validate incoming policy and configure default policy requests.
+    void CPUActivityAgent::validate_policy(std::vector<double> &in_policy) const
+    {
+        GEOPM_DEBUG_ASSERT(in_policy.size() == M_NUM_POLICY,
+                           "CPUActivityAgent::" + std::string(__func__) +
+                           "(): policy vector not correctly sized.  Expected  " +
+                            std::to_string(M_NUM_POLICY) + ", actual: " +
+                            std::to_string(in_policy.size()));
+
+        ///////////////////////
+        //CPU POLICY CHECKING//
+        ///////////////////////
+        double freq_core_min = m_platform_io.read_signal("CPU_FREQUENCY_MIN_AVAIL", GEOPM_DOMAIN_BOARD, 0);
+        double freq_core_max = m_platform_io.read_signal("CPU_FREQUENCY_MAX_AVAIL", GEOPM_DOMAIN_BOARD, 0);
```

Comment:
```validate_policy()``` gets called every time the Controller walks down.  Having these individual ```read_signal()``` calls in here will slow this down.  Do you need to read them every time through, or just once?

If it's just once, I'd put these values in a member variable, and cache them in ``init()`` like you do for the controls.